File: leetcode2/0543. Diameter of Binary Tree.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def diameterOfBinaryTree(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        if not root:
            return 0
        
        root=self.remarktree(root)
        root_list=[root]
        rep=0
        while root_list:
            len_root_list=len(root_list)
            for _ in range(len_root_list):
                cur_root=root_list.pop()
                if cur_root.right and cur_root.left:
                    rep=max(rep,cur_root.left.val+cur_root.right.val)
                    root_list.append(cur_root.right)
                    root_list.append(cur_root.left)
                elif not cur_root.right and cur_root.left:
                    rep=max(rep,cur_root.left.val)
                    root_list.append(cur_root.left)
                elif cur_root.right and not cur_root.left:
                    rep=max(rep,cur_root.right.val)
                    root_list.append(cur_root.right)
        return rep
                
        
        
    def remarktree(self,root):
        if not root.left and not root.right:
            root.val=1
        elif not root.left and root.right:
            root.val=1+self.remarktree(root.right).val
        elif root.left and not root.right:
            root.val=1+self.remarktree(root.left).val
        else:
            root.val=1+max(self.remarktree(root.left).val,+self.remarktree(root.right).val)
        return root

File: C#/0543. Diameter of Binary Tree.cs
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     public int val;
 *     public TreeNode left;
 *     public TreeNode right;
 *     public TreeNode(int x) { val = x; }
 * }
 */
public class Solution {
    public int DiameterOfBinaryTree(TreeNode root) {
        if (root==null){
            return 0;
        }
        int rep=0;
        rep=Math.Max(rep,DiameterOfBinaryTree(root.left));
        rep=Math.Max(rep,DiameterOfBinaryTree(root.right)); 
        rep=Math.Max(rep,CalHeight(root.left)+CalHeight(root.right));
        return rep;
    }
    
    public int CalHeight(TreeNode root){
        int rep=0;
        if (root==null){
            return rep;
        }
        return 1+Math.Max(CalHeight(root.left),CalHeight(root.right));
    }
}
